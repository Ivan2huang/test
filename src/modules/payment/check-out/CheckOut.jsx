import React, { useEffect, useCallback, useContext } from 'react';
import { SingletonRouter } from 'next/router';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import Head from 'next/head';

import { Box, Card, CardContent, withStyles } from '@material-ui/core';

import CheckOutMethodFormContainer from '../CheckOutMethodFormContainer';
import Loading from '../Loading';
import Grid from '../../../uiComponents/Grid';
import GridItem from '../../../uiComponents/GridItem';
import { orderType } from '../../../types';
import { IntlContext } from '../../../i18n/withIntlProvider';
import { isValidLanguageCode } from '../../../helpers/helpers';
import { getCookie } from '../../../helpers/auth';
import { LANG } from '../../../constants/auth';
import CONFIG from '../../../constants/config';

const styles = theme => ({
  card: {
    marginLeft: -30,
    paddingLeft: 30,
    [theme.breakpoints.up('md')]: {
      paddingRight: 30,
    },
  },
});

const CheckOut = ({
  router: { query },
  getPaymentMethods,
  checkOutOrder,
  goBack,
  callbackUrl,
  order,
  isLoadingPaymentMethods,
  isCheckingOut,
  classes,
  mpgsSessionUrl,
}) => {
  const { amount, currency } = order;
  const setLocale = useContext(IntlContext);

  useEffect(() => {
    const lang = getCookie(LANG);
    const locale = isValidLanguageCode(lang) ? lang : CONFIG.defaultLanguage;

    setLocale(locale);
  }, []);

  useEffect(() => {
    getPaymentMethods({
      clientId: query.query.client_id,
      userId: query.query.user_id,
      amount,
      currency,
      callbackUrl,
    });
  }, [query.query.client_id, query.query.user_id, amount, currency]);

  const onCheckOutOrder = useCallback(
    postOrder => {
      checkOutOrder(
        postOrder,
        query.query.user_id,
        query.query.client_id,
        callbackUrl,
      );
    },
    [query.query.user_id, query.query.client_id, callbackUrl],
  );

  const goBackToCart = () => {
    goBack(callbackUrl);
  };

  return (
    <>
      <Head>
        <script type="text/javascript" src={mpgsSessionUrl} />
      </Head>
      <Box>
        <Card className={classes.card} data-testid="checkout-wrapper">
          <CardContent>
            <Box mt={10} mb={10}>
              <Grid>
                <GridItem columns={{ xs: 12, md: 12 }}>
                  <Box md={{ xs: 12 }}>
                    {isCheckingOut ? (
                      <Loading />
                    ) : (
                      <CheckOutMethodFormContainer
                        checkOutOrder={onCheckOutOrder}
                        isLoadingPaymentMethods={isLoadingPaymentMethods}
                        order={order}
                        goBack={goBackToCart}
                      />
                    )}
                  </Box>
                </GridItem>
              </Grid>
            </Box>
          </CardContent>
        </Card>
      </Box>
    </>
  );
};

CheckOut.propTypes = {
  router: PropTypes.shape(SingletonRouter).isRequired,
  getPaymentMethods: PropTypes.func.isRequired,
  checkOutOrder: PropTypes.func.isRequired,
  goBack: PropTypes.func.isRequired,
  callbackUrl: PropTypes.string.isRequired,
  isLoadingPaymentMethods: PropTypes.bool.isRequired,
  isCheckingOut: PropTypes.bool.isRequired,
  order: orderType.isRequired,
  classes: PropTypes.shape({
    card: PropTypes.string.isRequired,
  }).isRequired,
  mpgsSessionUrl: PropTypes.string.isRequired,
};

export default injectIntl(withStyles(styles)(CheckOut));
