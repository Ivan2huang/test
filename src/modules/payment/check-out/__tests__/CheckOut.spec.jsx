import React from 'react';

import { render, cleanup } from '@testing-library/react';
import 'jest-dom/extend-expect';

import CheckOut from '../CheckOut';
import withTheme from '../../../../themes/withThemeProvider';
import withIntl from '../../../../i18n/withIntlProvider';
import withRedux from '../../../../redux/withReduxProvider';

jest.mock(
  '../../CheckOutMethodFormContainer',
  // eslint-disable-next-line react/prop-types
  () => ({ children, props }) => (
    <div {...props}>
      Dummy CheckOutMethodFormContainer Component
      {children}
    </div>
  ),
);

describe('CheckOut Component', () => {
  let props;
  const Component = withRedux(withTheme(withIntl(CheckOut)));

  beforeEach(() => {
    props = {
      router: {
        query: {
          query: {
            user_id: '3',
            client_id: '3',
          },
        },
      },
      getPaymentMethods: jest.fn(),
      checkOutOrder: jest.fn(),
      callbackUrl: 'https://httpbin.org',
      isLoadingPaymentMethods: false,
      isCheckingOut: false,
      goBack: jest.fn(),
      order: {
        merchantOrderId: '001',
        amount: 100,
        currency: '$',
        redirectUrl: 'https://httpbin.org',
        instrumentId: 3,
        sessionId: 'XXX001',
        orderItems: [
          {
            name: 'ORDER-1',
            description: 'DESCRIPTION',
            sku: 'SKU-1',
            qty: 1,
            unitPrice: 100,
            unit: 'UNIT-1',
          },
        ],
      },
      mpgsSessionUrl: 'http://localhost',
    };

    window.PaymentSession = {};
    window.PaymentSession.configure = jest.fn().mockReturnValue({});
    window.PaymentSession.updateSessionFromForm = jest.fn();
  });

  afterEach(cleanup);

  it('should match the snapshot with loading circle', () => {
    props.isCheckingOut = true;
    const { container } = render(<Component {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot with dummy checkout form', () => {
    props.isCheckingOut = false;
    const { container } = render(<Component {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should call get payment methods', () => {
    const postParams = {
      clientId: props.router.query.query.client_id,
      userId: props.router.query.query.user_id,
      amount: props.order.amount,
      currency: props.order.currency,
      callbackUrl: 'https://httpbin.org',
    };

    render(<Component {...props} />);

    expect(props.getPaymentMethods).toHaveBeenCalledTimes(1);
    expect(props.getPaymentMethods).toHaveBeenCalledWith(postParams);
  });
});
