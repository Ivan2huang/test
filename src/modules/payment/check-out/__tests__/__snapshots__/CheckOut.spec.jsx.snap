// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckOut Component should match the snapshot with dummy checkout form 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-125"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root CheckOut-card-123 MuiPaper-rounded"
      data-testid="checkout-wrapper"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiBox-root MuiBox-root-155"
        >
          <div
            class="MuiGrid-root MuiGrid-container"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CheckOut Component should match the snapshot with loading circle 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-3"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root CheckOut-card-1 MuiPaper-rounded"
      data-testid="checkout-wrapper"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiBox-root MuiBox-root-33"
        >
          <div
            class="MuiGrid-root MuiGrid-container"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
